File: electron/backend/shared/middleware/auth.middleware.js
import * as AuthService from '../../modules/auth/auth.service.js';

/**
 * Middleware to verify session token
 */
export function authenticate(req, res, next) {
    if (req.path.startsWith('/auth/')) {
        return next();
    }

    const authHeader = req.headers['authorization'];
    if (!authHeader) {
        return res.status(401).json({ error: 'Unauthorized: No token provided' });
    }

    const token = authHeader.split(' ')[1];
    if (!token) {
        return res.status(401).json({ error: 'Unauthorized: Invalid token format' });
    }

    const session = AuthService.getSession(token);
    if (!session) {
        return res.status(401).json({ error: 'Unauthorized: Session expired or invalid' });
    }

    req.user = session.user;
    req.permissions = session.permissions;
    req.token = token;

    next();
}

/**
 * Middleware to require specific permission
 * Supports single string or array of strings (OR logic)
 */
export function requirePermission(permission) {
    return (req, res, next) => {
        if (!req.user || !req.permissions) {
            return res.status(401).json({ error: 'Unauthorized: User not authenticated' });
        }

        const userPermissions = req.permissions;
        let hasAccess = false;

        if (Array.isArray(permission)) {
            hasAccess = permission.some(p => userPermissions.includes(p));
        } else {
            hasAccess = userPermissions.includes(permission);
        }

        if (!hasAccess) {
            return res.status(403).json({ error: `Forbidden: Missing required permission (${Array.isArray(permission) ? permission.join(' OR ') : permission})` });
        }

        next();
    };
}

File: src/main/java/com/possum/application/auth/AuthMiddleware.java
package com.possum.application.auth;

import com.possum.domain.exceptions.AuthenticationException;
import com.possum.domain.exceptions.AuthorizationException;

import java.util.List;

public class AuthMiddleware {

    private final AuthService authService;
    private final AuthorizationService authorizationService;

    public AuthMiddleware(AuthService authService, AuthorizationService authorizationService) {
        this.authService = authService;
        this.authorizationService = authorizationService;
    }

    public AuthUser authenticate(String authHeader) {
        if (authHeader == null || authHeader.isBlank()) {
            throw new AuthenticationException("Unauthorized: No token provided");
        }

        String[] parts = authHeader.split(" ");
        if (parts.length != 2 || !"Bearer".equals(parts[0])) {
            throw new AuthenticationException("Unauthorized: Invalid token format");
        }

        String token = parts[1];
        AuthUser user = authService.validateSession(token);

        if (user == null) {
            throw new AuthenticationException("Unauthorized: Session expired or invalid");
        }

        return user;
    }

    public void requirePermission(AuthUser user, String permission) {
        UserContext context = new UserContext(user.id(), user.roles(), user.permissions());
        
        if (!authorizationService.hasPermission(context, permission)) {
            throw new AuthorizationException("Forbidden: Missing required permission (" + permission + ")");
        }
    }

    public void requireAnyPermission(AuthUser user, List<String> permissions) {
        UserContext context = new UserContext(user.id(), user.roles(), user.permissions());
        
        if (!authorizationService.hasAnyPermission(context, permissions)) {
            throw new AuthorizationException("Forbidden: Missing required permission (" + String.join(" OR ", permissions) + ")");
        }
    }

    public void requireRole(AuthUser user, String role) {
        UserContext context = new UserContext(user.id(), user.roles(), user.permissions());
        
        if (!authorizationService.hasRole(context, role)) {
            throw new AuthorizationException("Forbidden: Missing required role (" + role + ")");
        }
    }

    public void requireAnyRole(AuthUser user, List<String> roles) {
        UserContext context = new UserContext(user.id(), user.roles(), user.permissions());
        
        if (!authorizationService.hasAnyRole(context, roles)) {
            throw new AuthorizationException("Forbidden: Missing required role (" + String.join(" OR ", roles) + ")");
        }
    }

    public void requireAdmin(AuthUser user) {
        UserContext context = new UserContext(user.id(), user.roles(), user.permissions());
        
        if (!authorizationService.isAdmin(context)) {
            throw new AuthorizationException("Forbidden: Admin access required");
        }
    }
}
